src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -152,12 +156,83 @@ class SetupPyChroot:
   setup_keywords_json: str
 
 
+@dataclass(frozen=True)
+class RunSetupPyRequest:
+  """A request to run a setup.py command."""
+  exported_target: ExportedTarget
+  chroot: SetupPyChroot
+  args: Tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class RunSetupPyResult:
+  """The result of running a setup.py command."""
+  output: Digest  # The state of the chroot after running setup.py.
+
+
 @dataclass(frozen=True)
 class SetuptoolsSetup:
   """The setuptools tool."""
   requirements_pex: Pex
 
 
+class SetupPyOptions(GoalSubsystem):
+  name = "setup-py2"
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--args', type=list, member_type=str,
+      help="Arguments to pass directly to setup.py, e.g. "
+           "`--setup-py2-args=\"bdist_wheel\"`. If unspecified, we just dump the setup.py chroot."
+    )
+
+
+class SetupPy(Goal):
+  """Runs setup.py commands."""
+  subsystem_cls = SetupPyOptions
+
+
+@console_rule
+async def run_setup_pys(addresses: BuildFileAddresses, console: Console, options: SetupPyOptions,
+                        distdir: DistDir, workspace: Workspace) -> SetupPy:
+  """Run setup.py commands on all exported targets addressed."""
+  targets = await Get[HydratedTargets](BuildFileAddresses, addresses)
+
+  exported_targets: List[ExportedTarget] = []
+  nonexported_targets: List[HydratedTarget] = []
+  for target in targets:
+    if _is_exported(target):
+      exported_targets.append(ExportedTarget(target))
+    else:
+      nonexported_targets.append(target)
+  if nonexported_targets:
+    raise TargetNotExported(
+      'Cannot run setup.py on these targets, because they have no `provides=` clause: '
+      f'{", ".join(so.address.reference() for so in nonexported_targets)}')
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
+    provides = target.hydrated_target.adaptor.provides
+    setup_py_dir = distdir.relpath / f'{provides.name}-{provides.version}'
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
+      setup_py_result = await Get[RunSetupPyResult](
+        RunSetupPyRequest(target, chroot, tuple(options.values.args)))
+      result_digest = setup_py_result.output
+    else:
+      # Just dump the chroot.
+      result_digest = chroot.digest
+    workspace.materialize_directory(
+      DirectoryToMaterialize(result_digest, path_prefix=str(setup_py_dir))
+    )
+    console.print_stderr(f'Wrote setup.py chroot for {addr} to {setup_py_dir}')
```

Comment:
Why stderr?

Reply:
+1. This feels like the actual result of the rule, and so should be stdout. And unlike the above, it _should_ probably go to the console, because it will/should happen every time the console_rule runs.

Reply:
I generally think of stdout as for the formal output of a thing, and stderr for logging/information. So a rule whose stated purpose is to output some data (e.g., `goals`) should do that to stdout. But this rule's purpose is not to output anything, this output is just incidental. For example, we could change its phrasing without incurring a deprecation cycle.

But probably it should just be logging?

Reply:
Naa, I tried logging and it doesn't look right. I think directly writing to stderr is the right thing here.